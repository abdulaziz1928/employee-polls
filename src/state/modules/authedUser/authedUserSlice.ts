import { PayloadAction, createSlice } from "@reduxjs/toolkit";

const initialState: string | null = null;

const authedUserSlice = createSlice({
  name: "authedUser",
  initialState: initialState as string | null,
  reducers: {
    setAuthedUser: (_, action: PayloadAction<string>) => action.payload,
    logoutAuthedUser: () => null,
  },
});

export const { setAuthedUser, logoutAuthedUser } = authedUserSlice.actions;

export default authedUserSlice.reducer;
