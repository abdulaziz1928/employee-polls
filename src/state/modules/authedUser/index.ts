import reducer from "./authedUserSlice";
export * from "./authedUserSlice";
export default reducer;
