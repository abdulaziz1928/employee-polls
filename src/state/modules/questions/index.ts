import reducer from "./questionsSlice";
export * from "./questionsSlice";
export * from "./models/question";
export default reducer;
