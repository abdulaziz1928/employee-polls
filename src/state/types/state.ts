import LoadingStatus from "./loading_status";

export default interface IState<T> {
  entities: T;
  loading: LoadingStatus;
  error?: string;
}
