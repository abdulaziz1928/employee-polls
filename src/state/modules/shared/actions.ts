import { Dispatch } from "@reduxjs/toolkit";
import { setAuthedUser } from "../authedUser";
import { getInitialData } from "../../apis/api";
import { receiveQuestions } from "../questions";
import { receiveUsers } from "../users";

const AUTHED_USER_ID = "tylermcginnis";

export function handleInitialData() {
  return (dispatch: Dispatch) => {
    return getInitialData().then(({ users, questions }) => {
      dispatch(receiveUsers(users));
      dispatch(receiveQuestions(questions));
      dispatch(setAuthedUser(AUTHED_USER_ID));
    });
  };
}
