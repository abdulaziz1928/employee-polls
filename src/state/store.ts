import authedUser from "./modules/authedUser";
import questions from "./modules/questions";
import users from "./modules/users";
import logger from "./middlewares/logger";
import prefrences from "./modules/prefrences/prefrencesSlice";
import { combineReducers, configureStore } from "@reduxjs/toolkit";

const reducers = { authedUser, users, questions, prefrences };

const rootReducer = combineReducers(reducers);

const middleWares = [logger];

export const store = configureStore({
  reducer: rootReducer,
  middleware: (gDM) => gDM().concat(...middleWares),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
