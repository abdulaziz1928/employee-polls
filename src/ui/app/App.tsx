import { useEffect } from "react";
import { useAppDispatch } from "./hooks";
import { handleInitialData } from "../../state/modules/shared/actions";
import { CssBaseline } from "@mui/material";
import { Routes, Route, Navigate } from "react-router-dom";
import PageRoutes from "../../state/types/page_routes";
import HomePage from "../features/home/home_page";
import LeaderboardPage from "../features/leaderboard/leaderboard_page";
import NewPollPage from "../features/new_poll/new_poll_page";
import PollPage from "../features/poll/poll_page";
import Login from "../features/login/login_page";
import PrivateRoutes from "../features/common/private_routes";
import SuccessSnackbar from "../features/common/success_snackbar";
import NotFoundPage from "../features/not_found/not_found_page";

export default function App() {
  const dispatch = useAppDispatch();
  useEffect(() => {
    dispatch(handleInitialData());
  }, [dispatch]);
  return (
    <>
      <CssBaseline />
      <SuccessSnackbar />
      <Routes>
        <Route element={<PrivateRoutes />}>
          <Route path={PageRoutes.Home} element={<HomePage />} />
          <Route path={PageRoutes.Leaderboard} element={<LeaderboardPage />} />
          <Route path={PageRoutes.Question} element={<PollPage />} />
          <Route path={PageRoutes.New} element={<NewPollPage />} />
          <Route path={PageRoutes.NotFound} element={<NotFoundPage />} />
          <Route path="*" element={<Navigate to={PageRoutes.NotFound} />} />
        </Route>
        <Route path={PageRoutes.Login} element={<Login />} />
      </Routes>
    </>
  );
}
