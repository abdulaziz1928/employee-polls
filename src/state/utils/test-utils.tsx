import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { setupStore, AppStore, RootState } from "../store";
import { MemoryRouter } from "react-router-dom";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithProvidersAndRouter(
  ui: React.ReactElement,

  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route: string = "/"
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <MemoryRouter initialEntries={[route]}>{children}</MemoryRouter>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
