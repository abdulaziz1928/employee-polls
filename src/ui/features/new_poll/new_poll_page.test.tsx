import { Route, Routes } from "react-router-dom";
import { AppStore, setupStore } from "../../../state/store";
import NewPollPage from "./new_poll_page";
import PageRoutes from "../../../state/types/page_routes";
import { renderWithProvidersAndRouter } from "../../../state/utils/test-utils";
import { setAuthedUser } from "../../../state/modules/authedUser";
import { handleInitialData } from "../../../state/modules/shared/actions";
import { fireEvent, screen, waitFor } from "@testing-library/react";
import * as hooks from "../../app/hooks";
import * as router from "react-router";
let store: AppStore;

const page = (
  <Routes>
    <Route path={PageRoutes.New} element={<NewPollPage />} />
  </Routes>
);

describe("NewPollPage", () => {
  beforeAll(async () => {
    store = setupStore();
    await store.dispatch(handleInitialData());
    store.dispatch(setAuthedUser("tylermcginnis"));
  });

  it("should render the component", () => {
    const { container } = renderWithProvidersAndRouter(<NewPollPage />);
    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it("should submit a new poll and navigate", async () => {
    renderWithProvidersAndRouter(page, { store }, PageRoutes.New);

    const optionOneTextField = screen.getByTestId("option-one-text-field");
    const optionTwoTextField = screen.getByTestId("option-two-text-field");
    const submitButton = screen.getByTestId("submit-poll-button");

    expect(optionOneTextField).toBeInTheDocument();
    expect(optionTwoTextField).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();

    const useDispatchMock = jest.spyOn(hooks, "useAppDispatch");
    const useNavigateMock = jest.spyOn(router, "useNavigate");

    fireEvent.change(optionOneTextField, {
      target: { value: "Be a java developer" },
    });
    fireEvent.change(optionTwoTextField, {
      target: { value: "Be a javascript developer" },
    });
    fireEvent.click(submitButton);

    await waitFor(() => expect(useDispatchMock).toBeCalled());
    expect(useNavigateMock).toBeCalled();
  });
});
