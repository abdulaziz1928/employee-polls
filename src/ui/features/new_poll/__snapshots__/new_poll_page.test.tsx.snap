// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPollPage should render the component 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd css-lfn7w7-MuiContainer-root"
  >
    <div
      class="MuiStack-root css-j378z1-MuiStack-root"
    >
      <div
        class="MuiBox-root css-0"
      >
        <h1
          class="MuiTypography-root MuiTypography-h3 MuiTypography-alignCenter css-47suzj-MuiTypography-root"
        >
          Would You Rather
        </h1>
        <h2
          class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter MuiTypography-gutterBottom css-m2bl7s-MuiTypography-root"
        >
          Create Your Own Poll
        </h2>
      </div>
      <form
        class="MuiBox-root css-0"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <div
            class="MuiStack-root css-pwj874-MuiStack-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="optionOne"
                id="optionOne-label"
              >
                Option One
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="option-one-text-field"
                  id="optionOne"
                  name="optionOne"
                  required=""
                  style="height: 0px; overflow: hidden;"
                />
                <textarea
                  aria-hidden="true"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  readonly=""
                  style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                  tabindex="-1"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Option One
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="optionTwo"
                id="optionTwo-label"
              >
                Option Two
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="option-two-text-field"
                  id="optionTwo"
                  name="optionTwo"
                  required=""
                  style="height: 0px; overflow: hidden;"
                />
                <textarea
                  aria-hidden="true"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  readonly=""
                  style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                  tabindex="-1"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Option Two
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1212rsq-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
            data-testid="submit-poll-button"
            id=":r2:"
            tabindex="0"
            type="submit"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Submit
            </h6>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
